File: openacc/cuda_c/acc_main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

extern void gpukernel_wrapper(int N, float *C, float *A, float *B);

int main(int argc, char **argv)
{
  int i, N, x;
  float *A, *B, *C;

  x = 10;
  N = 128;
  A = (float*)malloc(sizeof(float)*N);
  B = (float*)malloc(sizeof(float)*N);
  C = (float*)malloc(sizeof(float)*N);

  for(i=0;i<N;i++){
	C[i] = 0.0f;	B[i] = 2.0f;
	A[i] = (float)(i+1)/(float)(N);
  }

  printf("A\n");
  for(i=0; i<N; i++){
	if(i%x==x-1){
	  printf(" %2.4f\n", A[i]);
	}else{
	  printf(" %2.4f", A[i]);
	}
  }
  printf("\n");
  printf("B\n");
  for(i=0; i<N; i++){
	if(i%x==x-1){
	  printf(" %2.4f\n", B[i]);
	}else{
	  printf(" %2.4f", B[i]);
	}
  }
  printf("\n");
  printf("C (before)\n");
  for(i=0; i<N; i++){
	if(i%x==x-1){
	  printf(" %2.4f\n", C[i]);
	}else{
	  printf(" %2.4f", C[i]);
	}
  }
  printf("\n");

#pragma acc enter data copyin(A[0:N],B[0:N],C[0:N])

#pragma acc kernels present(A,B,C)
#pragma acc loop independent
	for(i=0; i<N; i++){
	  C[i] += A[i] * B[i];
	}

#pragma acc host_data use_device(A,B,C)
	{
	  gpukernel_wrapper(N, C, A, B);
	}

#pragma acc exit data copyout(C[0:N])

  printf("C (after)\n");
  for(i=0; i<N; i++){
	if(i%x==x-1){
	  printf(" %2.4f\n", C[i]);
	}else{
	  printf(" %2.4f", C[i]);
	}
  }
  printf("\n");

  free(A); free(B); free(C);
  return 0;
}

File: openacc/cuda_f/acc_main.f90
program main
  use gpukernel
  implicit none

!  real(kind=4), allocatable, dimension(:) :: A, B, C
  real(4), allocatable, dimension(:) :: A, B, C
  integer :: I, N, X

  X = 10
  N = 128
  allocate(A(N), B(N), C(N))

  C = 0.0;  B = 2.0
  do I=1, N
     A(I) = real(I)/real(N)
  enddo

  write(*,*) "A"
  do I=1, N
     if(mod(I,X).eq.0)then
        write(*,'(F8.4)') A(I)
     else
        write(*,'(F8.4)',advance='no') A(I)
     endif
  enddo
  write(*,*) ""
  write(*,*) "B"
  do I=1, N
     if(mod(I,X).eq.0)then
        write(*,'(F8.4)') B(I)
     else
        write(*,'(F8.4)',advance='no') B(I)
     endif
  enddo
  write(*,*) ""
  write(*,*) "C (before)"
  do I=1, N
     if(mod(I,X).eq.0)then
        write(*,'(F8.4)') C(I)
     else
        write(*,'(F8.4)',advance='no') C(I)
     endif
  enddo
  write(*,*) ""


!$acc enter data copyin(A(1:N), B(1:N), C(1:N))

!$acc kernels
!$acc loop
  do I=1, N
     C(I) = C(I) + A(I) * B(I)
  enddo
!$acc end kernels

!$acc host_data use_device(A, B, C)
  call gpukernel_wrapper(N, C, A, B)
!$acc end host_data

!$acc exit data copyout(C(1:N))

  write(*,*) "C (after)"
  do I=1, N
     if(mod(I,X).eq.0)then
        write(*,'(F8.4)') C(I)
     else
        write(*,'(F8.4)',advance='no') C(I)
     endif
  enddo
  write(*,*) ""

  deallocate(A, B, C)

end program main
